### DISCUSSION TITLE
Overriding NEXTAUTH_URL during runtime for multitenancy

### DISCUSSION THREAD

**Discussion by rajat1saxena** [nextauthjs/next-auth#8154] [Help]
### Question 💬

Per this [reply](https://github.com/nextauthjs/next-auth/issues/600#issuecomment-1560551411), I can override `NEXTAUTH_URL` dynamically and I've tested the same in my multitenant app and it is working correctly.

Someone in the same thread mentioned that this approach isn't thread safe. Is it really a concern here and is this a good way to accommodate multi-tenancy since it is not supported officially as of now?

### How to reproduce ☕️

N/A

### Contributing 🙌🏽

Yes, I am willing to help answer this question in a PR

**Comment by ZainKhalidOfficial** [comment]
Did you find any solution?

**Comment by aman-godara** [comment]
short answer: No, it is not a good way to override environment variables in your code.
Overriding environment variables through code is never a good way of doing anything (whether you are doing it in next-auth or not).

So use this solution at your own risk.

**Comment by ZainKhalidOfficial** [comment]
Short Answer: It's not recommended, and multitenancy is not supported. 

Some devs may find a work around but if you don't know what you're doing then please look for an alternative option. I wasted a lot of time looking for it and, in the end, had to look for an alternative option.

Clerk Auth supports it otherwise go for a custom auth solution.

**Comment by swve** [comment]
Hey 👋🏻, if you're still struggling with this issue, here's an update : 

- With NextAuth v5, the NEXTAUTH_URL (host) now is fetched from the request headers, and you no longer need to put a fixed url : https://authjs.dev/getting-started/deployment#auth_url 
- They also pushed a fix for Oauth Providers when your users need to login from multiple subdomains or preview domains or completely different domains : https://authjs.dev/getting-started/deployment#securing-a-preview-deployment 
- Here is guide on how to migrate to v5 : https://authjs.dev/getting-started/migrating-to-v5 

I hope it fixes your issues.

**Comment by FDiskas** [comment]
Possible fix https://github.com/nextauthjs/next-auth/pull/13365

**Comment by ali3825** [comment]
---

Overriding `process.env.NEXTAUTH_URL` at runtime is not thread-safe because environment variables are global to the Node.js process. In concurrent environments (especially serverless), mutating it per request can cause race conditions or cross-tenant leakage.

If you're using older versions, this approach works but is risky and not officially supported.

In NextAuth v5 (Auth.js), this is no longer necessary — the host is inferred from the incoming request headers, which makes multi-domain and preview deployments work without dynamically overriding `NEXTAUTH_URL`.

So the recommended solution is to migrate to v5 instead of modifying environment variables at runtime.

Docs:

* [https://authjs.dev/getting-started/deployment#auth_url](https://authjs.dev/getting-started/deployment#auth_url)
* [https://authjs.dev/getting-started/migrating-to-v5](https://authjs.dev/getting-started/migrating-to-v5)

---

**Reply by FDiskas** [reply]
Just letting you all know that resolving host is not fully working - I forced to use a patch https://github.com/nextauthjs/next-auth/pull/13365
